detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Detektor.kt

```diff
@@ -70,12 +76,41 @@ class Detektor(
         return awaitAll(tasks).filterNotNull()
     }
 
-    private fun KtFile.analyze(bindingContext: BindingContext): Map<RuleSetId, List<Finding>> =
-        providers.asSequence()
+    private fun analyze(file: KtFile, bindingContext: BindingContext): Map<RuleSetId, List<Finding>> {
+        val ruleSets = providers.asSequence()
             .filter { it.isActive(config) }
             .map { it.createRuleSet(config) }
-            .filter { it.shouldAnalyzeFile(this, config) }
-            .sortedBy { it.id }
-            .map { ruleSet -> ruleSet.id to ruleSet.visitFile(this, bindingContext) }
-            .toMergedMap()
+            .filter { it.shouldAnalyzeFile(file, config) }
+            .associate { it.id to it.rules }
+
+        fun isCorrectable(rule: BaseRule): Boolean = when (rule) {
+            is Rule -> rule.autoCorrect
+            is MultiRule -> rule.rules.any { it.autoCorrect }
+            else -> error("No other rule type expected.")
+        }
+
+        val (correctableRules, otherRules) =
+            ruleSets.asSequence()
+                .flatMap { (_, value) -> value.asSequence() }
+                .partition { isCorrectable(it) }
+
+        val result = HashMap<RuleSetId, MutableList<Finding>>()
+
+        fun executeRules(rules: List<BaseRule>) {
+            for (rule in rules) {
+                rule.visitFile(file, bindingContext)
+                for (finding in rule.findings) {
+                    val mappedRuleSet = idMapping[finding.id]
+                        ?: error("Mapping for '${finding.id}' expected.")
+                    result.putIfAbsent(mappedRuleSet, ArrayList())
+                    result[mappedRuleSet]?.add(finding)
```

Comment:
```suggestion
                    result[mappedRuleSet]!!.add(finding)
```

This case should never happen but it's better to crash that miss some finding.